Java/core/src/main/java/com/amazon/randomcutforest/preprocessor/ImputePreprocessor.java

```diff
@@ -0,0 +1,404 @@
+                    getShingledInput(shingleSize - 1), deviations, clipFactor);
+            // note that initial values are all interpolated by 0,fixed, or linear
+            // there are no missing values to handle
+            updateState(initialValues[i], scaledInput, initialTimeStamps[i], lastInputTimeStamp, null);
+            if (forest.isInternalShinglingEnabled()) {
+                // updateAllowed = not updateShingleOnly
+                forest.update(scaledInput, !updateAllowed());
+            } else {
+                if (updateAllowed()) {
+                    forest.update(lastShingledPoint);
+                }
+            }
+        }
+        initialTimeStamps = null;
+        initialValues = null;
+    }
+
+    /**
+     * determines the gap between the last known timestamp and the current timestamp
+     * 
+     * @param timestampGap current gap
+     * @param averageGap   the average gap (often determined by
+     *                     timeStampDeviation.getMean()
+     * @return the number of positions till timestamp
+     */
+    protected int determineGap(long timestampGap, double averageGap) {
+        if (internalTimeStamp <= 1) {
+            return 1;
+        } else {
+            double gap = timestampGap / averageGap;
+            return (gap >= 1.5) ? (int) Math.ceil(gap) : 1;
+        }
+    }
+
+    public int numberOfImputes(long timestamp) {
+        long lastInputTimeStamp = previousTimeStamps[shingleSize - 1];
+        return determineGap(timestamp - lastInputTimeStamp, getTimeFactor(timeStampDeviations[1])) - 1;
+    }
+
+    /**
+     * a single function that constructs the next shingle, with the option of
+     * committing them to the forest However the shingle needs to be generated
+     * before we process a point; and can only be committed once the point has been
+     * scored. Having the same deterministic transformation can be useful. Note for
+     * this imputation timestamp cannot be missing
+     *
+     * @param averageGap   the gap in timestamps
+     * @param changeForest boolean determining if we commit to the forest or not
+     * @param forest       the resident RCF
+     * @return the next shingle
+     */
+    protected float[] generateShingle(double[] inputTuple, long timestamp, int[] missingValues, double averageGap,
+            boolean changeForest, RandomCutForest forest) {
+        long lastInputTimeStamp = previousTimeStamps[shingleSize - 1];
+        double[] input = Arrays.copyOf(inputTuple, inputLength);
+        double[] previous = getShingledInput(shingleSize - 1);
+        double[] savedInput = Arrays.copyOf(previous, inputLength);
+        int numberToImpute = determineGap(timestamp - lastInputTimeStamp, averageGap) - 1;
+
+        if (imputationMethod != RCF || !forest.isOutputReady()) {
+            ImputationMethod method = (imputationMethod == RCF) ? DEFAULT_DYNAMIC : imputationMethod;
```

Comment:
Oops - fixed. method should be used below (for imputationMethod because we are covering for the case the forest is not yet ready -- and thus we cannot have RCF imputation).